src/Payum/Core/CoreGatewayFactory.php

```diff
@@ -89,7 +122,13 @@ public function createConfig(array $config = array())
                 }
 
                 if (class_exists(HttpCurlClient::class)) {
-                    return new HttpCurlClient();
+                    if ($config['httplug.message_factory'] instanceof \Closure) {
+                        $config['httplug.message_factory'] = $config['httplug.message_factory']->__invoke($config);
```

Comment:
make sure `httplug.stream_factory` and `httplug.message_factory` are added here prior http client option.

https://github.com/Payum/Payum/blob/master/src/Payum/Core/CoreGatewayFactory.php#L170